src/fonduer/utils/utils_udf.py

```diff
@@ -0,0 +1,158 @@
+import logging
+
+from scipy.sparse import csr_matrix
+from sqlalchemy.dialects.postgresql import insert
+
+from fonduer.candidates.models import Candidate
+
+logger = logging.getLogger(__name__)
+
+# Flag value to signal that no filtering on split should be applied. Not an
+# integer to ensure that it won't conflict with a user's split value.
+ALL_SPLITS = "ALL"
+
+
+def _get_cand_values(candidate, key_table):
+    """Get the corresponding values for the key_table."""
+    # NOTE: Import just before checking to avoid circular imports.
+    from fonduer.features.models import FeatureKey
+    from fonduer.supervision.models import GoldLabelKey, LabelKey
+
+    if key_table == FeatureKey:
+        return candidate.features
+    elif key_table == LabelKey:
+        return candidate.labels
+    elif key_table == GoldLabelKey:
+        return candidate.gold_labels
+    else:
+        raise ValueError("{} is not a valid key table.".format(key_table))
+
+
+def batch_upsert_records(session, table, records):
+    """Batch upsert records into postgresql database."""
+    if not records:
+        return
+
+    stmt = insert(table.__table__).values(records)
+    stmt = stmt.on_conflict_do_update(
+        constraint=table.__table__.primary_key,
+        set_={"keys": stmt.excluded.get("keys"), "values": stmt.excluded.get("values")},
+    )
+    session.execute(stmt)
+    session.commit()
+
+
+def get_sparse_matrix(session, key_table, cand_lists, key=None):
```

Comment:
Beside this, we probably need a function call get_keys which returns all the keys in the sparse matrix in order.

Reply:
Ok, which classes should expose that function? Labeler and Featurizer?